pkg/storage/util.go

```diff
@@ -106,23 +104,19 @@ func NamespaceKeyFunc(prefix string, obj runtime.Object) (string, error) {
 	if err != nil {
 		return "", err
 	}
-	name := meta.GetName()
-	if msgs := validation.IsValidPathSegmentName(name); len(msgs) != 0 {
-		return "", fmt.Errorf("invalid name: %v", msgs)
-	}
-	return prefix + "/" + meta.GetNamespace() + "/" + name, nil
+	// We assume that NamespaceKeyFunc is used for read requests (List and Watch)
+	// and thus the object is correct and we don't perform any validation.
+	return prefix + "/" + meta.GetNamespace() + "/" + meta.GetName(), nil
 }
 
 func NoNamespaceKeyFunc(prefix string, obj runtime.Object) (string, error) {
 	meta, err := meta.Accessor(obj)
 	if err != nil {
 		return "", err
 	}
-	name := meta.GetName()
-	if msgs := validation.IsValidPathSegmentName(name); len(msgs) != 0 {
```

Comment:
I don't agree these are "most matches".
Most watches will always come from kubelet and kube-proxy, because there are tens/hundreds or even thousands of them, whereas there is only constant number of those coming from controllers. 

Anyway - I agree with the second argument, that we shouldn't have two implementations of the same thing. But that would be much bigger change (as well as computing a key only once). So we won't be able to do it for 1.4 (at least noone in Google will be).